File: calibration_scripts/stereoCalibration.py
import cv2
import numpy as np
import time

from calibrationTools import *
from serial import Serial
from picamera2 import Picamera2

def main():
    
    imageSize = (1200, 720)
    numPoses = 41
    
    serialPort = Serial("/dev/ttyAMA0", 450)
    
    camL = Picamera2(0)
    camR = Picamera2(1)
    
    config = camL.create_video_configuration(main={"format": 'RGB888', "size": imageSize})
    camL.configure(config)
    
    config = camR.create_video_configuration(main={"format": 'RGB888', "size": imageSize})
    camR.configure(config)
    
    # camL.sensor_modes.append({"format": "RGB888"})
    # camL.sensor_modes.append({"size": imageSize})
    # camL.sensor_modes.append({"fps": 120})
    
    # camR.sensor_modes.append({"format": "RGB888"})
    # camR.sensor_modes.append({"size": imageSize})
    # camR.sensor_modes.append({"fps": 120})
    
    camL.start()
    camR.start()
    
    objp = createObjPoints(6, 8)
    
    for i in range(8):
        serialPort.write("\x00".encode())
        time.sleep(0.05)
        
    leftCornerList = []
    rightCornerList = []
    objPoints = []
        
    i = 0
    while i < numPoses:
        print("Capturing image: " + str(i))
        currentNum = 2
        keyPress = None
        # while keyPress != ord('a'):
        start = time.time()
        while currentNum > -1:
            serialPort.write("\x00".encode())
            leftIm = camL.capture_array()
            rightIm = camR.capture_array()
            
            vis = np.concatenate((leftIm, rightIm), axis=1)
            # vis = cv2.cvtColor(vis, cv2.COLOR_BGR2RGB)
            vis = cv2.putText(vis, str(currentNum), (0,60), cv2.FONT_HERSHEY_PLAIN, 4, (0, 255, 0), 5)
            vis = cv2.putText(vis, f"{i+1}/{numPoses}", (1270, 60), cv2.FONT_HERSHEY_PLAIN, 4, (0, 255, 0), 5)
            vis = cv2.resize(vis, (int(vis.shape[1] / 1.75), int(vis.shape[0] / 1.75)))
            cv2.imshow("Stereo Image", vis)
            keyPress = cv2.waitKey(10)
            if time.time() - start >= 1.0:
                currentNum -= 1
                start = time.time()
        leftGray = cv2.cvtColor(leftIm, cv2.COLOR_BGR2GRAY)
        rightGray = cv2.cvtColor(rightIm, cv2.COLOR_BGR2GRAY)
        leftFound, leftCorners = cv2.findChessboardCornersSB(leftGray, (6, 8))
        rightFound, rightCorners = cv2.findChessboardCornersSB(rightGray, (6, 8))
        if leftFound and rightFound:
            objPoints.append(objp)
            leftCornerList.append(leftCorners)
            rightCornerList.append(rightCorners)
            cv2.drawChessboardCorners(leftIm, (6, 8), leftCorners, leftFound)
            cv2.drawChessboardCorners(rightIm, (6, 8), rightCorners, rightFound)
            vis = np.concatenate((leftIm, rightIm), axis=1)
            vis = cv2.resize(vis, (int(vis.shape[1] / 1.75), int(vis.shape[0] / 1.75)))
            # vis = cv2.cvtColor(vis, cv2.COLOR_BGR2RGB)
            start = time.time()
            while time.time() - start < 2:
                cv2.imshow("Stereo Image", vis)
                cv2.waitKey(1)
            i += 1
        else:
            if leftFound:
                print("Could not find right chessboard")
            elif rightFound:
                print("Could not find left chessboard")
            else:
                print("Could not find any chessboard")
                
    leftIntrFile = cv2.FileStorage("/home/pi/Cam0Intr.xml", cv2.FILE_STORAGE_READ)
    rightIntrFile = cv2.FileStorage("/home/pi/Cam1Intr.xml", cv2.FILE_STORAGE_READ)
    camMatL = leftIntrFile.getNode("Matrix").mat()
    camDCL = leftIntrFile.getNode("DistCoeffs").mat()
    camMatR = rightIntrFile.getNode("Matrix").mat()
    camDCR = rightIntrFile.getNode("DistCoeffs").mat()
    
    flags = 0
    flags |= cv2.CALIB_FIX_INTRINSIC
    # Here we fix the intrinsic camara matrixes so that only Rot, Trns, Emat and Fmat are calculated.
    # Hence intrinsic parameters are the same 
    
    criteria_stereo= (cv2.TERM_CRITERIA_EPS + cv2.TERM_CRITERIA_MAX_ITER, 30, 0.001)
    
    
    # This step is performed to transformation between the two cameras and calculate Essential and Fundamenatl matrix
    retS, camMatL, camDCL, camMatR, camDCR, Rot, Trns, Emat, Fmat = cv2.stereoCalibrate(objPoints, leftCornerList, rightCornerList, camMatL, camDCL, camMatR, camDCR, imageSize, criteria_stereo, flags)
    
    if retS:
        print("Stereo calibrated")
    else:
        print("Stereo calibration failed")
        SystemExit()
    
    rect_l, rect_r, proj_mat_l, proj_mat_r, Q, roiL, roiR = cv2.stereoRectify(camMatL, camDCL, camMatR, camDCR, imageSize, Rot, Trns, 1,(0,0))
    
    Left_Stereo_Map = cv2.initUndistortRectifyMap(camMatL, camDCL, rect_l, proj_mat_l,
                                             imageSize, cv2.CV_16SC2)
    Right_Stereo_Map = cv2.initUndistortRectifyMap(camMatR, camDCR, rect_r, proj_mat_r,
                                                imageSize, cv2.CV_16SC2)
    
    print("Saving parameters ......")
    cv_file = cv2.FileStorage("/home/pi/StereoMaps.xml", cv2.FILE_STORAGE_WRITE)
    cv_file.write("Left_Stereo_Map_x",Left_Stereo_Map[0])
    cv_file.write("Left_Stereo_Map_y",Left_Stereo_Map[1])
    cv_file.write("Right_Stereo_Map_x",Right_Stereo_Map[0])
    cv_file.write("Right_Stereo_Map_y",Right_Stereo_Map[1])
    cv_file.release()
    
if __name__ == "__main__":
    main()

File: calibration_scripts/stereoCalibration.cpp

#include "calibration/stereoCalibration.hpp"

#include "lccv.hpp"
#include "tools/CanmoreImageTransmitter.hpp"
#include "canmore/client_ids.h"

#include <chrono>
#include <net/if.h>
#include <filesystem>
#include <opencv2/core/persistence.hpp>
#include <opencv2/opencv.hpp>
#include <unistd.h>

void sendSerial() {
    serialPort.Send('\x00');
}

void createCamera(lccv::PiCamera &camera_, int id) {
    camera_.options->camera = id;
    camera_.options->video_width = imageSize.width;
    camera_.options->video_height = imageSize.height;
    camera_.options->framerate = 60;
    // camera_.options->verbose = true;
}

int main(int argc, char *argv[]) {
    lccv::PiCamera camL, camR;

    createCamera(camL, 0);
    createCamera(camR, 1);

    camL.startVideo();
    camR.startVideo();

    bool useCan = false;

    if (argc >= 2) {
        useCan = std::stoi(argv[1]);
        if (useCan == 1) {
            printf("Displaying image over can\n");
            useCan = true;
        }
        else {
            printf("Displaying image over X server\n");
        }
    } else {
            printf("Displaying image over X server\n");
    }

    CanmoreImageTransmitter *imageTx;

    if (useCan) {
        int ifIdx = if_nametoindex("can0");
        if (!ifIdx) {
            throw std::system_error(errno, std::generic_category(), "if_nametoindex");
        }
        imageTx = new CanmoreImageTransmitter(ifIdx, CANMORE_CLIENT_ID_DOWNWARDS_CAMERA, 190, 20, true);
    }

    std::string saveFoldername;

    // Need to fill up buffer (fairly sure why we need to do this, not entirely sure lmao)
    for (int i = 0; i < 8; i++) {
        sendSerial();
    }

    int i = 0;
    cv::Mat leftIm, rightIm, vis;
    std::vector<std::vector<cv::Point2f>> leftCornersVector, rightCornersVector;
    std::vector<std::vector<cv::Point3f>> objPts;

    std::vector<cv::Point3f> objp;
    for (int i = 0; i < 8; i++) {
        for (int j = 0; j < 6; j++) {
            objp.push_back(cv::Point3f(j, i, 0));
        }
    }

    std::stringstream stream;
    std::time_t current_time = std::time(nullptr);
    std::tm tm = *std::localtime(&current_time);
    stream << std::put_time(&tm, "%Y%m%d%H%M");
    saveFoldername = "/home/pi/CalibrationImgs/" + stream.str() + "_StereoImgs/";
    std::filesystem::create_directories(saveFoldername.c_str());

    int keyPress = 0;

    while (i < totalImages) {
        int currentNum = 1;
        auto start = std::chrono::high_resolution_clock::now();
        //while (keyPress != 'c') {
        while (currentNum > -1) {
            sendSerial();
            if (!camL.getVideoFrame(leftIm, 99999999) || !camR.getVideoFrame(rightIm, 99999999)) {
                printf("Cannot grab video frame\n");
            }
            cv::cvtColor(leftIm, leftIm, cv::COLOR_RGB2GRAY);
            cv::cvtColor(rightIm, rightIm, cv::COLOR_RGB2GRAY);
            cv::hconcat(leftIm, rightIm, vis);
            cv::putText(vis, std::to_string(currentNum), cv::Point2d(60, 0), cv::FONT_HERSHEY_PLAIN, 4,
                        cv::Scalar(0, 255, 0), 5);
            std::string displayText = std::to_string(i + 1) + "/" + std::to_string(totalImages);
            cv::putText(vis, displayText, cv::Point(60, leftIm.rows - 200), cv::FONT_HERSHEY_PLAIN, 4,
                        cv::Scalar(0, 255, 0), 5);
	    cv::cvtColor(vis, vis, cv::COLOR_GRAY2RGB);
            // cv::resize(vis, vis, cv::Size(vis.cols / 1.75, vis.rows / 1.75));
            if (useCan) {
                imageTx->transmitImage(vis);
                usleep(200000);
                keyPress = imageTx->getKeypress();
            } else {
                cv::imshow("Calibration", vis);
                keyPress = cv::waitKey(1) & 255;
            }
            if (std::chrono::duration_cast<std::chrono::seconds>(std::chrono::high_resolution_clock::now() - start)
                    .count() >= 1) {
                currentNum -= 1;
                start = std::chrono::high_resolution_clock::now();
            }
        }
        //char *filename = new char[100];
        //sprintf(filename, "/home/pi/202404161747_StereoImgs/Left_%i.png", i);
        //leftIm = cv::imread(filename);
        //filename = new char[100];
        //sprintf(filename, "/home/pi/202404161747_StereoImgs/Right_%i.png", i);
        //rightIm = cv::imread(filename);

        std::vector<cv::Point2f> leftCorners, rightCorners;
        bool leftFound = cv::findChessboardCornersSB(leftIm, patternSize, leftCorners);
        bool rightFound = cv::findChessboardCornersSB(rightIm, patternSize, rightCorners);
        if (leftFound && rightFound) {
            objPts.push_back(objp);
            leftCornersVector.push_back(leftCorners);
            rightCornersVector.push_back(rightCorners);
            cv::imwrite(saveFoldername + "Left_" + std::to_string(i) + ".png", leftIm);
            cv::imwrite(saveFoldername + "Right_" + std::to_string(i) + ".png", rightIm);
            cv::drawChessboardCorners(leftIm, patternSize, leftCorners, leftFound);
            cv::drawChessboardCorners(rightIm, patternSize, rightCorners, rightFound);
            cv::hconcat(leftIm, rightIm, vis);
            start = std::chrono::high_resolution_clock::now();
            while (std::chrono::duration_cast<std::chrono::seconds>(std::chrono::high_resolution_clock::now() - start)
                       .count() < 0) {
                // Just show the checkerboard
                cv::imshow("Calibration", vis);
                cv::waitKey(1);
            }
            i++;
        }
        else {
            if (leftFound) {
                printf("Could not find right chessboard\n");
            }
            else if (rightFound) {
                printf("Could not find left chessboard\n");
            }
            else {
                printf("Could not find any chessboard\n");
            }
        }
        keyPress = 0;
    }

    // camL.stopVideo();
    // camR.stopVideo();
    
    cv::FileStorage leftIntrFile("/home/pi/Cam0Intr.xml", cv::FileStorage::READ);
    cv::FileStorage rightIntrFile("/home/pi/Cam1Intr.xml", cv::FileStorage::READ);

    cv::Mat camMatL, camMatR, camDCL, camDCR;
    leftIntrFile["Matrix"] >> camMatL;
    leftIntrFile["DistCoeffs"] >> camDCL;
    rightIntrFile["Matrix"] >> camMatR;
    rightIntrFile["DistCoeffs"] >> camDCR;
    leftIntrFile.release();
    rightIntrFile.release();

    int flags = 0 | cv::CALIB_FIX_INTRINSIC;

    cv::Mat R, T, E, F;
    cv::TermCriteria criteria(cv::TermCriteria::EPS + cv::TermCriteria::MAX_ITER, 30, 0.001);
    double retS = cv::stereoCalibrate(objPts, leftCornersVector, rightCornersVector, camMatL, camDCL, camMatR, camDCR,
                                      imageSize, R, T, E, F, flags, criteria);
    if (retS) {
        printf("Stereo Calibrated\n");
    }
    else {
        printf("Stereo Calibration Failed\n");
    }

    cv::Mat rectL, rectR, projL, projR, Q, roiL, roiR;
    cv::stereoRectify(camMatL, camDCL, camMatR, camDCR, leftIm.size(), R, T, rectL, rectR, projL, projR, Q, 1);
    cv::Mat stereoMapLX, stereoMapLY, stereoMapRX, stereoMapRY;
    cv::initUndistortRectifyMap(camMatL, camDCL, rectL, projL, imageSize, CV_16SC2, stereoMapLX, stereoMapLY);
    cv::initUndistortRectifyMap(camMatR, camDCR, rectR, projR, imageSize, CV_16SC2, stereoMapRX, stereoMapRY);
    // printf("%d\n", stereoMapRY.type());

    cv::FileStorage cvFile = cv::FileStorage("/home/pi/StereoMaps.xml", cv::FileStorage::WRITE);
    cvFile.write("Left_Stereo_Map_x", stereoMapLX);
    cvFile.write("Left_Stereo_Map_y", stereoMapLY);
    cvFile.write("Right_Stereo_Map_x", stereoMapRX);
    cvFile.write("Right_Stereo_Map_y", stereoMapRY);
    cvFile.release();
}
